src/renderers/dom/shared/__tests__/ReactDOMComponent-test.js

```diff
@@ -618,6 +618,22 @@ describe('ReactDOMComponent', function() {
       expect(SimpleEventPlugin.didPutListener.mock.calls.length).toBe(2);
       expect(SimpleEventPlugin.willDeleteListener.mock.calls.length).toBe(1);
     });
+
+    it('should warn for children on void elements', function() {
+      spyOn(console, 'error');
+      var X = React.createClass({
+        render: function() {
+          return <input>moo</input>;
+        }
```

Comment:
On a related note, I'd kind of like to add an exception for objects and arrays nested immediately within another object or array. It's quite common to have an array of a single element, and to end up with syntax like:

```js
var config = {
  rules: [{
    foo: true,
    bar: false,
  }],
}
```